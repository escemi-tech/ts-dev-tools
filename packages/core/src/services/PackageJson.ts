import { PackageJson as PackageJsonType } from "type-fest";

import { PackageJsonMerge } from "./PackageJsonMerge";
import { FileService } from "./FileService";
import { join } from "path";

export type JsonArray = boolean[] | number[] | string[] | JsonFileData[] | Date[];
export type AnyJson = boolean | number | string | JsonFileData | Date | JsonArray | JsonArray[];

export type JsonFileData = {
  [key: string]: AnyJson | undefined;
};
export type PackageJsonContent = JsonFileData & PackageJsonType;

export class PackageJson {
  private static readonly PACKAGE_JSON_FILE_NAME = "package.json";

  private content?: PackageJsonContent = undefined;

  constructor(private path: string) {
    if (!FileService.fileExists(this.path)) {
      throw new Error(`Package.json "${this.path}" does not exist`);
    }
  }

  getPath(): string {
    return this.path;
  }

  getContent(): PackageJsonContent {
    if (this.content) {
      return this.content;
    }
    return (this.content = JSON.parse(FileService.getFileContent(this.path)) as PackageJsonContent);
  }

  setContent(content: PackageJsonContent): void {
    this.content = content;
    this.write();
  }

  getPackageName(): string | undefined {
    return this.getContent().name;
  }

  getPackageVersion(): string | undefined {
    return this.getContent().version;
  }

  isPrivate(): boolean {
    return this.getContent().private === true;
  }

  getTsDevToolsVersion(): string | undefined {
    const tsDevToolsConfig = this.getContent().tsDevTools as JsonFileData | undefined;
    const version = tsDevToolsConfig?.version as string | undefined;
    return version;
  }

  getDependenciesPackageNames(): string[] {
    const dependencies = this.getContent().dependencies;
    return dependencies ? Object.keys(dependencies) : [];
  }

  getDevDependenciesPackageNames(): string[] {
    const devDependencies = this.getContent().devDependencies;
    return devDependencies ? Object.keys(devDependencies) : [];
  }

  getAllDependenciesPackageNames(): string[] {
    return Array.from(
      new Set([...this.getDependenciesPackageNames(), ...this.getDevDependenciesPackageNames()])
    );
  }

  merge(update: PackageJsonContent): void {
    this.content = PackageJsonMerge.merge(this.getContent(), update);
    this.write();
  }

  backup(): string {
    const backupPath = this.path + ".backup";
    FileService.copyFile(this.path, backupPath);
    return backupPath;
  }

  restore(backupPath: string): void {
    FileService.copyFile(backupPath, this.path);
    this.content = undefined;
  }

  private write() {
    FileService.putFileContent(this.path, JSON.stringify(this.content, null, 2));
  }

  static fromDirPath(dirPath: string): PackageJson {
    const packageJsonPath = join(dirPath, PackageJson.PACKAGE_JSON_FILE_NAME);
    return new PackageJson(packageJsonPath);
  }
}
