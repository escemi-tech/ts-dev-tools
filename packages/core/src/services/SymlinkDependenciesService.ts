import { existsSync, symlinkSync } from "fs";
import { join } from "path";

import { PackageJson } from "./PackageJson";
import { Plugin, PluginService } from "./PluginService";

export class SymlinkDependenciesService {
  static DEPENDENCIES_FOLDER = "node_modules";

  static executeSymlinking(absoluteProjectDir: string): void {
    console.info(`Symlinking dev dependencies...`);

    const installedPlugins = PluginService.getInstalledPlugins(absoluteProjectDir);

    for (const plugin of installedPlugins) {
      SymlinkDependenciesService.symlinkPluginDependencies(absoluteProjectDir, plugin);
    }

    console.info(`Symlinking dev dependencies done!`);
  }

  private static symlinkPluginDependencies(absoluteProjectDir: string, plugin: Plugin) {
    const pluginDependencies = SymlinkDependenciesService.getPluginDependencies(plugin);

    for (const pluginDependency of pluginDependencies) {
      const pluginDependencyPath = join(
        plugin.path,
        SymlinkDependenciesService.DEPENDENCIES_FOLDER,
        pluginDependency
      );

      if (!existsSync(pluginDependencyPath)) {
        continue;
      }

      const projectPluginDependencyPath = join(
        absoluteProjectDir,
        SymlinkDependenciesService.DEPENDENCIES_FOLDER,
        pluginDependency
      );
      if (existsSync(projectPluginDependencyPath)) {
        continue;
      }

      console.info(`- Symlinking ${pluginDependency}`);
      SymlinkDependenciesService.symlinkDependency(
        pluginDependencyPath,
        projectPluginDependencyPath
      );
    }
  }

  private static getPluginDependencies(plugin: Plugin): string[] {
    const pluginPackageJson = PackageJson.fromDirPath(plugin.path);
    return pluginPackageJson.getDependenciesPackageNames();
  }

  private static symlinkDependency(
    pluginDependencyPath: string,
    projectPluginDependencyPath: string
  ): void {
    symlinkSync(pluginDependencyPath, projectPluginDependencyPath);
  }
}
