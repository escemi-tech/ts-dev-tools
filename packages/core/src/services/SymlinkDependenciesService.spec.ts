import { getConsoleInfoContent, mockConsoleInfo, resetMockedConsoleInfo } from "../tests/console";
import {
  createTestProjectDirWithFixtures,
  removeTestProjectDir,
  restorePackageJson,
} from "../tests/project";
import { PackageJson } from "./PackageJson";
import { SymlinkDependenciesService } from "./SymlinkDependenciesService";

describe("SymlinkDependenciesService", () => {
  let testProjectDir: string;

  beforeAll(() => {
    testProjectDir = createTestProjectDirWithFixtures(__filename);
  });

  afterAll(() => {
    removeTestProjectDir(__filename);
  });

  describe("executeSymlinking", () => {
    beforeEach(() => {
      mockConsoleInfo();
    });

    afterEach(() => {
      resetMockedConsoleInfo();
      restorePackageJson(__filename);
    });

    it("should symlink dependencies", () => {
      PackageJson.fromDirPath(testProjectDir).merge({
        devDependencies: {
          "@ts-dev-tools/core": "1.0.0",
        },
      });

      const executeSymlinkingAction = () =>
        SymlinkDependenciesService.executeSymlinking(testProjectDir);

      expect(executeSymlinkingAction).not.toThrow();
      expect(getConsoleInfoContent()).toMatchSnapshot();
    });
  });
});
