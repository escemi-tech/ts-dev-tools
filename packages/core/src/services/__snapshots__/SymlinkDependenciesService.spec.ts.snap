// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SymlinkDependenciesService executeSymlinking should symlink dependencies 1`] = `
"Symlinking dev dependencies...
- Symlinking @types/node
Symlinking dev dependencies done!"
`;
